src/coreclr/tools/Common/Compiler/TypeExtensions.cs

```diff
@@ -182,43 +182,68 @@ public static bool IsArrayTypeWithoutGenericInterfaces(this TypeDesc type)
             {
-                result = type1 == type2;
+                return type1 == type2;
             }
+
             // If either or both types are canonical subtypes, we can sometimes prove inequality.
-            else
+            if (AreGuaranteedToRepresentDifferentTypes(type1, type2))
+            {
+                return false;
+            }
+
+            return null;
+
+            static bool AreGuaranteedToRepresentDifferentTypes(TypeDesc type1, TypeDesc type2)
             {
-                // If either is a value type then the types cannot
-                // be equal unless the type defs are the same.
-                if (type1.IsValueType || type2.IsValueType)
+                if (type1.IsCanonicalDefinitionType(CanonicalFormKind.Any) || type2.IsCanonicalDefinitionType(CanonicalFormKind.Any))
                 {
-                    if (!type1.IsCanonicalDefinitionType(CanonicalFormKind.Universal) && !type2.IsCanonicalDefinitionType(CanonicalFormKind.Universal))
-                    {
-                        if (!type1.HasSameTypeDefinition(type2))
-                        {
-                            result = false;
-                        }
-                    }
+                    // Universal canonical definition can match any type. We can't prove inequality.
+                    if (type1.IsCanonicalDefinitionType(CanonicalFormKind.Universal) || type2.IsCanonicalDefinitionType(CanonicalFormKind.Universal))
+                        return false;
+
+                    return type1.IsGCPointer != type2.IsGCPointer;
                 }
-                // If we have two ref types that are not __Canon, then the
-                // types cannot be equal unless the type defs are the same.
-                else
+
+                TypeFlags category = type1.Category;
+                if (category != type2.Category)
+                    return true;
+
+                switch (category)
                 {
-                    if (!type1.IsCanonicalDefinitionType(CanonicalFormKind.Any) && !type2.IsCanonicalDefinitionType(CanonicalFormKind.Any))
-                    {
-                        if (!type1.HasSameTypeDefinition(type2))
+                    case TypeFlags.Array:
+                        if (((ArrayType)type1).Rank != ((ArrayType)type2).Rank)
+                            return true;
+                        return AreGuaranteedToRepresentDifferentTypes(((ArrayType)type1).ElementType, ((ArrayType)type2).ElementType);
+                    case TypeFlags.SzArray:
+                    case TypeFlags.ByRef:
+                    case TypeFlags.Pointer:
+                        return AreGuaranteedToRepresentDifferentTypes(((ParameterizedType)type1).ParameterType, ((ParameterizedType)type2).ParameterType);
+
+                    default:
+                        if (type1.IsDefType || type2.IsDefType)
```

Comment:
I think this could get away with the same check as the VM side.

```suggestion
                        if (type1.IsDefType)
```

Reply:
The VM side does check of matching CorElementType as early out that guarantees that both or neither types are def types.

I do not think that matching `TypeFlags` gives us the same guarantee here. The managed type system is more flexible, so I think we should check both types.